test/GenerationSandbox.Unmarshalled.Tests/ComHelpers.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if NET7_0_OR_GREATER
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Windows.Win32.Foundation;
+using Windows.Win32.System.Com;
+
+namespace Windows.Win32;
+
+// The `unsafe` modifier is only allowed to appear on the class declaration -- not the partial method declaration.
+// See https://github.com/dotnet/csharplang/discussions/7298 for more.
+internal unsafe partial class ComHelpers
+{
+    static partial void PopulateIUnknownImpl<TComInterface>(IUnknown.Vtbl* vtable)
+        where TComInterface : unmanaged
+    {
+        // IUnknown member initialization of the v-table would go here.
+        vtable->QueryInterface_1 = (delegate* unmanaged[Stdcall]<IUnknown*, Guid*, void**, HRESULT>)&QueryInterface;
+        vtable->AddRef_2 = (delegate* unmanaged[Stdcall]<IUnknown*, uint>)&AddRef;
+        vtable->Release_3 = (delegate* unmanaged[Stdcall]<IUnknown*, uint>)&Release;
+    }
+
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvStdcall) })]
+    private static HRESULT QueryInterface(IUnknown* punk, Guid* iid, void** ppvObject)
+    {
+        throw new NotImplementedException();
+    }
+
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvStdcall) })]
+    private static uint AddRef(IUnknown* punk)
+    {
+        throw new NotImplementedException();
+    }
+
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvStdcall) })]
+    private static uint Release(IUnknown* punk)
+    {
+        throw new NotImplementedException();
+    }
```

Comment:
@jeremykuhne For purposes of a mini-sample, I'd like to actually implement this properly (with ComWrappers) if possible. But I couldn't figure out how. Can you help?

Reply:
Sure. To just get it to populate the IUnknown entries you can do this:

``` C#
public class TestComWrappers : ComWrappers
{
    private static readonly IUnknown.Vtbl* s_wrappersUnknown = GetComWrappersUnknown();

    private static IUnknown.Vtbl* GetComWrappersUnknown()
    {
        IUnknown.Vtbl* vtable = (IUnknown.Vtbl*)RuntimeHelpers.AllocateTypeAssociatedMemory(
            typeof(CustomComWrappers),
            sizeof(IUnknown.Vtbl));

        GetIUnknownImpl(out nint fpQueryInterface, out nint fpAddRef, out nint fpRelease);
        vtable->QueryInterface_1 = (delegate* unmanaged[Stdcall]<IUnknown*, Guid*, void**, HRESULT>)fpQueryInterface;
        vtable->AddRef_2 = (delegate* unmanaged[Stdcall]<IUnknown*, uint>)fpAddRef;
        vtable->Release_3 = (delegate* unmanaged[Stdcall]<IUnknown*, uint>)fpRelease;

        return vtable;
    }

    // Call this to populate the Vtbl
    internal static void PopulateIUnknown(IUnknown.Vtbl* vtable)
    {
        vtable->QueryInterface_1 = s_wrappersUnknown->QueryInterface_1;
        vtable->AddRef_2 = s_wrappersUnknown->AddRef_2;
        vtable->Release_3 = s_wrappersUnknown->Release_3;
    }

    // Abstracts that need implementation
    protected override unsafe ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count)
    {
        count = 0;
        return null;
    }

    protected override object? CreateObject(nint externalComObject, CreateObjectFlags flags) => null;
    protected override void ReleaseObjects(IEnumerable objects) => throw new NotImplementedException();
}
```

Creating a full CCW requires implementing `ComputeVtables`.